File: src/main/java/io/trydent/olimpo/http/HttpRoute.java
package io.trydent.olimpo.http;

import io.vertx.ext.web.Router;

import java.util.function.UnaryOperator;

import static io.trydent.olimpo.http.HttpHeader.Value.ApplicationJson;

public interface HttpRoute extends UnaryOperator<Router> {
  static HttpRoute webroot(String path, HttpExchange exchange) {
    return new WebrootRoute(path, exchange);
  }

  static HttpRoute action(String path, HttpExchange exchange) {
    return new ActionRoute(path, exchange);
  }
}

final class WebrootRoute implements HttpRoute {
  private final String path;
  private final HttpExchange exchange;

  WebrootRoute(final String path, final HttpExchange exchange) {
    this.path = path;
    this.exchange = exchange;
  }

  @Override
  public final Router apply(final Router router) {
    router.get(path).handler(exchange.get());
    return router;
  }
}

final class ActionRoute implements HttpRoute {
  private final String path;
  private final HttpExchange exchange;

  ActionRoute(final String path, final HttpExchange exchange) {
    this.path = path;
    this.exchange = exchange;
  }

  @Override
  public final Router apply(final Router router) {
    router.post(path)
      .consumes(ApplicationJson)
      .produces(ApplicationJson)
      .handler(exchange.get());
    return router;
  }
}

File: src/main/java/io/trydent/olimpo/http/HttpRoute.kt
package io.trydent.olimpo.http

import io.trydent.olimpo.vertx.HttpValue.ApplicationJson
import io.trydent.olimpo.vertx.consumes
import io.trydent.olimpo.vertx.produces
import io.vertx.ext.web.Router

interface HttpRoute : (Router) -> Router {
  companion object {
    fun webroot(path: String, exchange: HttpExchange): HttpRoute = WebrootRoute(path, exchange)
    fun action(path: String, exchange: HttpExchange): HttpRoute = ActionRoute(path, exchange)
  }
}

internal class WebrootRoute(private val path: String, private val exchange: HttpExchange) : HttpRoute {
  override fun invoke(router: Router) = router.apply {
    get(path).handler(exchange())
  }
}

internal class ActionRoute(private val path: String, private val exchange: HttpExchange) : HttpRoute {
  override fun invoke(router: Router): Router = router.apply {
    post(path)
      .consumes(ApplicationJson)
      .produces(ApplicationJson)
      .handler(exchange())
  }
}
